src/fs/patch.ts

```diff
@@ -1,398 +1,96 @@
-import { Stats } from 'fs'
+import { getLibzipSync } from '@yarnpkg/libzip'
+import { patchFs, npath, PosixFS, NodeFS, ZipFS } from '@yarnpkg/fslib'
+import { SnapshotZipFS } from './SnapshotZipFS'
+import * as assert from 'assert'
 
-export interface NexeBinary {
+export interface NexeHeader {
   blobPath: string
-  resources: { [key: string]: number[] }
   layout: {
-    stat: Stats
     resourceStart: number
-    contentSize?: number
-    contentStart?: number
-    resourceSize?: number
+    resourceSize: number
+    contentSize: number
+    contentStart: number
   }
 }
 
 let originalFsMethods: any = null
 let lazyRestoreFs = () => {}
+const patches = (process as any).nexe.patches || {}
+delete (process as any).nexe
 
-// optional Win32 file namespace prefix followed by drive letter and colon
-const windowsFullPathRegex = /^(\\{2}\?\\)?([a-zA-Z]):/
-
-const upcaseDriveLetter = (s: string): string =>
-  s.replace(windowsFullPathRegex, (_match, ns, drive) => `${ns || ''}${drive.toUpperCase()}:`)
-
-function shimFs(binary: NexeBinary, fs: any = require('fs')) {
+function shimFs(binary: NexeHeader, fs: typeof import('fs') = require('fs')) {
   if (originalFsMethods !== null) {
     return
   }
+
   originalFsMethods = Object.assign({}, fs)
-  const { blobPath, resources: manifest } = binary,
-    { resourceStart, stat } = binary.layout,
-    directories: { [key: string]: { [key: string]: boolean } } = {},
-    notAFile = '!@#$%^&*',
-    isWin = process.platform.startsWith('win'),
-    isString = (x: any): x is string => typeof x === 'string' || x instanceof String,
-    noop = () => {},
-    path = require('path'),
-    winPath: (key: string) => string = isWin ? upcaseDriveLetter : s => s,
-    baseDir = winPath(path.dirname(process.execPath))
+
+  const realFs: typeof fs = { ...fs }
+  const nodeFs = new NodeFS(realFs)
+
+  const blob = Buffer.allocUnsafe(binary.layout.resourceSize)
+  const blobFd = realFs.openSync(binary.blobPath, 'r')
+  const bytesRead = realFs.readSync(
```

Comment:
I'm concerned that there is too much being allocated here. Currently, the blobs remain on disk and aren't loaded into memory until read.. Do you know if there is anyway to base this off of a file/offset so we can keep the snapshot on disk?

Reply:
re: looks like ZipFS would need to support accepting an offset, which it does not currently.

Reply:
IIRC at the time I was reasoning that for common cases it needed to be pulled into memory anyway, so if we bite the bullet and do that up front, it'll be faster.

Do zips require an offset?  They're specifically designed to keep the table of contents at the end of the file, and I think all contents are referenced by relative offset from the table of contents.  I think a lot of self-extracting installers are a pre-built exe appended with a zip.

I did a quick test on my Linux box:
```
wget zip-file-from-somewhere.zip
cp `which bash` ./exe
cat zip-file-from-somewhere.zip >> ./exe
./exe -c 'echo foo' # works like bash
unzip -vl ./exe # also appears to be a valid zip file
```

Reply:
You're right that that is generally the common case. I don't have very much evidence but am hesitant to lose the advantage of direct file reads via the offset currently used.

Reply:
I think you may not need an offset to do direct file reads, due to the way zip files are structured.